crates/boot_srv/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+#![deny(missing_docs)]
+//! Kitsune2 boot server is an HTTP REST server for handling bootstrapping
+//! discovery of peer network reachability in p2p applications.
+//!
+//! Despite being in the kitsune2 repo, `boot_srv` and `boot_cli` do not depend
+//! on any Kitsune2 crates. This is to ensure the bootstrapping functionality
+//! is well-defined, self-contained, easily testable in isolation, and
+//! usable for projects that don't choose to make use of Kitsune2 itself.
+//!
+//! That being said, the boot server and client are designed to transfer the
+//! `AgentInfoSigned` data type defined in the `kitsune2_api` crate. Since
+//! the canonical encoding of that type is JSON, we just redefine a subset
+//! of the schema here, and only validate the parts required for bootstrapping.
+//!
+//! For additional details, please see the [spec].
+
+/// This is a documentation module containing the kitsune2_boot spec.
+///
+/// #### 1. Types
+///
+/// - `Base64Agent` - base64UrlNoPad safe encoded string agent id.
+/// - `Base64Space` - base64UrlNoPad safe encoded string space id.
+/// - `Base64Sig` - base64UrlNoPad safe encoded string crypto signature.
+/// - `Json` - string containing json that can be decoded.
+/// - `I64` - string contaning an i64 number.
+///
+/// ```text
+/// AgentInfoSigned = { "agentInfo": Json, "signature": Base64Sig }
+/// AgentInfo = {
+///   "agent": Base64Agent,
+///   "space": Base64Space,
+///   "createdAt": I64,
+///   "expiresAt": I64,
+///   "isTombstone": boolean
+/// }
+/// ```
+///
+/// Any other properties on these objects will be ignored and pass through.
+///
+/// #### 2. REST API
+///
+/// ##### 2.1. In Brief
+///
+/// ```text
+/// ErrResponse = { "error": string }
+/// OkResponse = {}
+/// ListResponse = [ AgentInfoSigned, .. ]
+/// ```
+///
+/// - `PUT /boot/Base64Space/Base64Agent`
+///   - Request Body: `AgentInfoSigned`
+///   - Response Body: `OkResponse | ErrResponse`
+/// - `GET /boot/Base64Space`
+///   - Response Body: `ListResponse | ErrResponse`
+/// - `GET /`
+///   - Response Body: `OkResponse | ErrResponse`
+///
+/// ##### 2.2. Publishing info to the boot server.
+///
+/// A `PUT` on `/boot/Base64Space/Base64Agent` with an `AgentInfoSigned`
+/// json object as the request body.
+///
+/// - The server MUST reject the request if the body is > 1024 bytes.
+/// - The server MUST reject the request if `createdAt` is not within
+///   3 minutes in either direction of the server time.
+/// - The server MUST reject the request if `expiresAt` is in the past.
+/// - The server MUST reject the request if `signature` is invalid vs
+///   the `agentInfo`.
+/// - If `isTombstone` is `true`, the server MUST delete any existing
+///   info being held.
+/// - If `isTombstone` is `false`, the server MAY begin storing the info.
```

Comment:
I'd find it clearer to split create and delete into different endpoints. Then we could also drop the expiresAt field for delete requests.

Reply:
The tombstone field is require for the functionality on the kitsune side. Since in the active network they cannot just delete an agent info--they might get the old one gossiped back to them again, so they need tombstone records. And having them be gossiped just the same as other infos saves complexity. Since we need that anyways, I don't want to implement a different method for the bootstrap server.

Reply:
Consider the verifiability aspect. We cannot let just anybody delete any record, so we'd have to come up with some alternate payload for the clients to sign anyways... Can we just use the existing functionality?

Reply:
Ahh, I get why it's been done this way now. We'd have to sign something to send to the delete anyway... Yeah okay, fair enough :)